File: examples/nodejs/mapper.js
#!/usr/bin/env node

var stdin = process.stdin;
var stdout = process.stdout;
var input = '';

stdin.setEncoding('utf8');
stdin.on('data', function(data) {
  if (data) {
    input += data;
    while (input.match(/\r?\n/)) {
      input = RegExp.rightContext;
      proc(RegExp.leftContext);
    }
  }
});

stdin.on('end', function() {
  if (input) {
    proc(input);
  }
});

function proc(line) {
    var parts = line.split("\t");
    key = parts[0];
    val = JSON.parse(parts[1]);
    
    exec(key, val);
};

var emit = function(key, val){
	stdout.write(key + "\t" + val + "\n");
}

function exec(key, val) {
    if('pages' in val)
    {
        for(i in val.pages)
        {
            emit(i, 1);
        }
    }
}

File: examples/go/mapper.go
package main

import (
    "os"
    "bufio"
    "encoding/json"
    "strings"
    "fmt"
)


func proc(key string, values map[string]interface{}) {
    pages := values["pages"].(map[string]interface{})

    for page, _ := range pages {
        fmt.Printf("%s\t%d\n", page, len(pages))
    }
}

func main() {
    r := bufio.NewReader(os.Stdin)

    for {
        var data interface{}

        line, _, _ := r.ReadLine()

        lines := strings.Split(string(line), "\t")

        if len(lines) < 2 {
            continue
        }

        key := lines[0]
        json.Unmarshal([]byte(lines[1]), &data)
        m := data.(map[string]interface{})

        proc(key, m)

        fmt.Printf("|next|\n")
    }
}
